keps/sig-auth/0000-authority-delegation.md

```diff
@@ -0,0 +1,259 @@
+with the request:
+
+```
+Authority-Selector: {“authorityReference”:{“kind”:”ClusterRole”,”name”:”viewer”}}
+```
+or:
+```
+Authority-Selector: {“authorityReference”:{“kind”:”AuthorityDelegation”,”uid”:”0000-0...”}}
+```
+
+The value of the selector is a `SelectedAuthority` object in the
+`authorization.k8s.io` API group:
+
+```go
+type SelectedAuthority struct {
+  metav1.TypeMeta
+  AuthorityReference AuthorityReference
+}
+
+type AuthorityReference struct {
+  metav1.TypeMeta
+  Name string
+  UID types.UID
+}
+```
+
+When a request is received by an API, the API would normally iterate over the
+set of authorizers and each authorizer would check all policies that would
+permit the action, searching for a policy that permits the action and is bound
+to the user.
+
+When a request is received by the API with a selected authority, the API would
+check the specific authorizer that supports the AuthorityReference kind, and
+the authorizer would check the request against the specific policy referenced
+in by the SelectedAuthority.
+
+### Persistent Authority Delegation
+
+#### AuthorityDelegation
+
+A new non-namespaced persistent object will be added to the
+`authorization.k8s.io` API group to store instances of authority delegation.
+
+```go
+type AuthoritiyDelegation struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // The user that created the delegation.
+  User Info
+  // The principal that is delegating authority.
+  Subject SubjectRef
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // defined above
+  Authority AuthorityReference
+  // If this delegation is created selecting a previously delegated authority,
+  // the delegation chain is recorded here.
+  DelegationHistory []AuthorityDelegationReference
+}
+
+type SubjectRef struct {
+  Kind SubjectKind
+  Name string
+  UID types.UID
+}
+
+const (
+  GroupKind SubjectKind = "Group"
+  UserKind SubjectKind = "User"
+)
+
+type AuthrityDelegationReference struct {
+  UID types.UID
+}
+```
+
+#### AuthorityDelegationPolicy
+
+A new policy will be introduced to implictly create `AuthorityDelegations` in
+reaction to API actions. The `AuthorityDelegationPolicy` will be a
+non-namespaced object in the `authorization.k8s.io` API group.
+
+```go
+type AuthorityDelegationPolicy struct {
+  // The resource attributes of an action to implicitly create a delegation for.
+  ResourceAttributes *ResourceAttributes
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // The permissions that are delegated.
+  Authority AuthorityReference
+}
+```
+
+When an API action matching an `AuthorityDelegationPolicy` is processed, an
+`AuthorityDelegation` is implicitly created before any objects are persisted.
+
+#### Example of a Cascading Delegation
+
+Suppose a situation where:
+
+1) End user Foobar creates a deployment.
+2) Deployment controller creates a replicaset in response.
+3) ReplicaSet controller creates a pod in response.
```

Comment:
I think it would be helpful to flesh out this example more. E.g. where the AuthorityDelegationPolicy comes into play.

Reply:
IIUC, it looks like this?

1. User Foobar creates a deployment
2. AuthorityDelegationPolicy `deployment` automatically creates an authority delegation (pseudo-code):
```
kind: AuthoritiyDelegation
metadata:
  name: foobar-deployment
spec:
  user: system:authority-delegation-controller
  subject: users/foobar
  delegates:
    - system:deployment-controller
    - system:replicaset-controller
  authority:
    kind: rbac.authorization.k8s.io/ClusterRole
    name: deployment  # I'm a bit confused here - does user Foobar need access to this role?
```
3. Deployment controller observes a new deployment object, ...
    i. I think there's a race here with the creation of the authority delegation?
    ii. How does the deployment controller know which authority to select for the the create replicaset request?